From 1662e2c2f4fb6f02e1656c00ffbc8cdc3f2e48aa Mon Sep 17 00:00:00 2001
From: John Howard <howardjohn@google.com>
Date: Sun, 20 Mar 2022 19:50:59 -0700
Subject: pilot: use correct Request.Start for Delta XDS (#38027)

https://github.com/istio/istio/pull/37800 changed which PushContext we
use; however, it had a bug which caused some rare test instability,
primarily in TestTraffic. In the initiate attempt at the change above, a
similar issue occurred, but with a much higher rate of error
(https://github.com/istio/istio/pull/37604).

Unfortunately, we forgot to apply the same fix to the Delta code path.
This PR does that, eliminating the last test flakes.

I ran this PR and `master` against the assertion tests 100s of times
overnight. Without this change it failed about 5% of the time; with it
0% (due to this issue - a few unrelated test flakes did occur).

Long term, this split code path with Delta and SotW is a risk. This was
a known risk, intentionally favoring possible code drift over
destabilizing the primary code (SotW). I think now that we are on Go
1.18 we can move over to generics fairly soon, and much of this risk
  will be mitigated.
---
 pilot/pkg/xds/ads.go   | 3 +++
 pilot/pkg/xds/delta.go | 5 ++++-
 2 files changed, 7 insertions(+), 1 deletion(-)

diff --git a/pilot/pkg/xds/ads.go b/pilot/pkg/xds/ads.go
index e2188a8e38..0b12ee6065 100644
--- a/pilot/pkg/xds/ads.go
+++ b/pilot/pkg/xds/ads.go
@@ -234,6 +234,9 @@ func (s *DiscoveryServer) processRequest(req *discovery.DiscoveryRequest, con *C
 	}
 
 	request.Reason = append(request.Reason, model.ProxyRequest)
+	// The usage of LastPushTime (rather than time.Now()), is critical here for correctness; This time
+	// is used by the XDS cache to determine if a entry is stale. If we use Now() with an old push context,
+	// we may end up overriding active cache entries with stale ones.
 	request.Start = con.proxy.LastPushTime
 	// SidecarScope for the proxy may not have been updated based on this pushContext.
 	// It can happen when `processRequest` comes after push context has been updated(s.initPushContext),
diff --git a/pilot/pkg/xds/delta.go b/pilot/pkg/xds/delta.go
index 0cfc5eba66..de10b93f30 100644
--- a/pilot/pkg/xds/delta.go
+++ b/pilot/pkg/xds/delta.go
@@ -310,7 +310,10 @@ func (s *DiscoveryServer) processDeltaRequest(req *discovery.DeltaDiscoveryReque
 	}
 
 	request.Reason = append(request.Reason, model.ProxyRequest)
-	request.Start = time.Now()
+	// The usage of LastPushTime (rather than time.Now()), is critical here for correctness; This time
+	// is used by the XDS cache to determine if a entry is stale. If we use Now() with an old push context,
+	// we may end up overriding active cache entries with stale ones.
+	request.Start = con.proxy.LastPushTime
 	// SidecarScope for the proxy may has not been updated based on this pushContext.
 	// It can happen when `processRequest` comes after push context has been updated(s.initPushContext),
 	// but before proxy's SidecarScope has been updated(s.updateProxy).
-- 
2.35.3

